### PR TITLE
add support for image history - should close #715

### LINKED ISSUE #715: Provide support for "docker history" command.
Hey looks like docker-java library has not support for `docker history image-name` command.
I would like to give PR if you are allowing me to merge it to your project .

### PR DESCRIPTION
**Pull request by grossjonas**
#715 - I felt so free :)

Adds base functionality for image history.

Tests are a bit weak, but similar to `ListImages*` - how could I improve them?

---

This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/719)

### CODE REVIEW

File: src/test/java/com/github/dockerjava/core/TestDockerCmdExecFactory.java [hunk 1]
```diff
@@ -1,62 +1,8 @@
 package com.github.dockerjava.core;
 
 import com.github.dockerjava.api.async.ResultCallback;
-import com.github.dockerjava.api.command.AttachContainerCmd;
+import com.github.dockerjava.api.command.*;
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/TestDockerCmdExecFactory.java]: no hash imports

Comment (grossjonas) [inline src/test/java/com/github/dockerjava/core/TestDockerCmdExecFactory.java]: sorry, that one slipped through. Should be fixed now.

**Comment by grossjonas** [comment]
I don't get it.

In 4 of 6 builds every thing is fine - the other 2 have authorization problems( ```Registration: "Missing password field"``` and ```Registration: "Wrong email format (it has to match \"[^@]+@[^@]+\\.[^@]+\")"```).

**Comment by KostyaSha** [comment]
Please annotate with API version fields and commands according to first appeared version.

**Comment by grossjonas** [comment]
The history functionality seems to be around since [Remote API Version 1.0](https://github.com/docker/docker/blob/v1.7.0/docs/reference/api/docker_remote_api_v1.0.md). ```RemoteApiVersion``` only refers to the [documentation hosted on the docker site](https://docs.docker.com/engine/reference/api/docker_remote_api/). So I did not annotate the command interface.

Did I catch all missing annotations?

**Comment by KostyaSha** [comment]
@grossjonas you can search not deeper than 1.17. Was all fields available in this version?

**Comment by grossjonas** [comment]
Sorry, I am not quite sure which class we are talking about. I guess it's [```History```](https://github.com/grossjonas/docker-java/blob/master/src/main/java/com/github/dockerjava/api/model/History.java).
```Id```, ```Created``` and ```CreatedBy``` were available in [1.17](https://github.com/docker/docker/tree/v1.7.0/docs/reference/api); ```Tags```,```Size``` and ```Comment```  appeared in [1.19](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.19/) and are now annotated accordingly.

Btw. [RemoteAPIVersion](https://github.com/docker-java/docker-java/blob/master/src/main/java/com/github/dockerjava/core/RemoteApiVersion.java)'s links to [1.16](http://docs.docker.com/engine/reference/api/docker_remote_api_v1.16/) and [1.17](http://docs.docker.com/engine/reference/api/docker_remote_api_v1.17/) are dead, but Docker has a Github repository with older version [here](https://github.com/docker/docker/tree/master/docs/reference/api)(even older versions up to 1.0 are available under the Tag [1.7.0](https://github.com/docker/docker/tree/v1.7.0/docs/reference/api)).

**Comment by KostyaSha** [comment]
Could you create such test for model file https://github.com/docker-java/docker-java/blob/419900d074e6d9754dcd5c6290d93c3350bab010/src/test/java/com/github/dockerjava/api/model/AuthConfigTest.java#L25 ?

**Comment by grossjonas** [comment]
Just a short feedback: I am working on it, but I probably need till Tuesday or Wednesday because of RL obligations.

**Comment by grossjonas** [comment]
I created a test for the model fields. In order to do so I also fixed a wrong parameter name in ```CommitCmdExec```s.
Shall I squash into two commits?

File: src/main/java/com/github/dockerjava/api/model/History.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    private Long created;
+
+    @JsonProperty("CreatedBy")
+    private String createdBy;
+
+    /**
+     * @since {@link RemoteApiVersion#VERSION_1_19}
+     */
+    @JsonProperty("Tags")
+    private String[] tags;
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/History.java]: Use List

**Comment by KostyaSha** [comment]
no need to squash

File: src/main/java/com/github/dockerjava/api/model/History.java (hunk 2 above)
Comment (grossjonas) [inline src/main/java/com/github/dockerjava/api/model/History.java]: change it - it's now a list

**Comment by grossjonas** [comment]
I'm not a native speaker - should I rename ```History``` to ```HistoryEntry```?

**Comment by KostyaSha** [comment]
> I'm not a native speaker - should I rename History to HistoryEntry?

If you think that it will be more exact, then rename. You can also look into docker/docker sources to get idea how they call such structures.

**Comment by KostyaSha** [comment]
> I created a test for the model fields. In order to do so I also fixed a wrong parameter name in both of the CommitCmdExecs.

Link was to serder test. You need dump json sample or (and) pick from docs to construct object and compare with result.

File: src/test/java/com/github/dockerjava/core/command/ListImageImageHistoryCmdImplTest.java [hunk 3]
```diff
@@ -32,7 +32,7 @@
 import static org.hamcrest.Matchers.nullValue;
 
 @Test(groups = "integration")
-public class ListImageHistoryCmdImplTest extends AbstractDockerClientTest {
+public class ListImageImageHistoryCmdImplTest extends AbstractDockerClientTest {
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/ListImageImageHistoryCmdImplTest.java]: imageimage? first variant was fine.

Comment (grossjonas) [inline src/test/java/com/github/dockerjava/core/command/ListImageImageHistoryCmdImplTest.java]: sorry - fixed it

**Comment by grossjonas** [comment]
I added the serder test. When seeing the link for the first time I though you just wanted me to check all fields. Thank you for your patience.

**Comment by grossjonas** [comment]
I just extended the *ExecTest. What else can I do to get this PR ready to be merged?

**Comment by grossjonas** [comment]
@KostyaSha Is anything left to do in order to get this PR merged?

**Comment by KostyaSha** [comment]
tests on 1.10.3 are failing

**Comment by grossjonas** [comment]
I've merge with master; seems like no more tests are failing ...

**Comment by KostyaSha** [comment]
Could you rebase and fixup/squash your commits?

**Pull request closed without merging**